### ISSUE TITLE
[Feature request]: Portable mode

### ISSUE THREAD

**Issue by auwsom** [flatpak/flatpak#4497]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

Hello, I made it to Flatpak after using Snaps (I didn't like because it heavily bloated my root partition). Then I found AppImages which are great because I can put them on a separate central partition to my various Linux VMs and partitions, reducing redundant space wastage. This was my strategy before on the Windows side and worked well before I transitioned over because of the immense size of basic Windows 10 VM at around 30GB. Also, the whole uninstall process is eliminated if I find a better application. They are quite large, so I found Flatpaks are moveable and also reference a core package between applications making them lighter.

Trying to figure out how the applications would respond to access from other VMs or partitions after installing the core packages, I saw a couple of GitHub issues asking for easier installation to other directories. I had simply used a symbolic link that worked thankfully to another partition. I had to correct the permissions to 755 in that directory (and not sure if I'll need to do that for every install). But remembered the user data need to be moved as well. I also see there is a not exactly simple workaround process for backing up with Flatpacks. I wonder if again this could be done simply with symbolic links. 

My use case for this would be to setup an application similar to how I do VMs. Have a base installation and then another for tinkering and customizing. Anything goes wrong and I revert. Good things get installed to the base. I did this with PortableApps very nicely, especially with things like browsers which I could have several copies of for specific uses and logins, all secured nicely in Veracrypt vaults which limit access if my system was compromised and lock automatically if the computer is stolen, turned off or storage disconnected. Sweet and simple. I managed to find a way to do this with Chrome on Ubuntu even which took a very long search to find someone who had achieved it with a complicated script, but that once is working, the directory is self-contained and can be cloned and modified, and backed up endlessly.

So I just found that AppImages have a Portable Mode that works similarly to PortableApps. The admin just has to create one or two directories named with the application name plus .config and/or .home. It would be great to see this functionality in Flatpaks. I've read you devs may have different goals (sandboxing?), but I think this would really widen your user base. Having a 'core' install similar to the Launcher/Updater in PortableApps could make the Flatpaks system agnostic, while lightweight. Yes, there would need to be compatibility checks, but an Updater could keep track of that. 

Well, I wanted to give this input because a Portable Mode for Flatpaks would seem the best of both portability and lightweightness. And give some background of the really useful use cases of backup and user profiles, and security, so easily accomplished with portable installations. 

Here are the related issues on GH and links:
https://github.com/flatpak/flatpak/issues/408
https://github.com/flatpak/flatpak/pull/4018
https://github.com/flatpak/flatpak/issues/2129
https://docs.appimage.org/user-guide/portable-mode.html
https://portableapps.com/

**Comment by auwsom** [comment]
Yes, that's what I meant. I'm wondering how difficult it is to combine the core packages together into a single downloadable file per OS. I realize it would be larger than installing only the necessary files per application, but it would allow further separation of the applications to a different partition, where they could be OS agnostic. The 'core' files could be stored separately in the case of cloned OSes or VMs to reduce total size for more efficient backing up OS and VM images. 

I was reminded yesterday of my main distaste for Snaps, as when repairing a GPT backup table, somehow the UUIDs of my partitions were changed. My OS partitions without Snap were able to drop to a maintenance shell so I could fix /etc/fstab. My main partition where I had uninstalled the Snaps but not the service continually hung on the service trying to start in the kernel boot messages. I had to sever the Snapd target symlinks to boot from a liveCD and then reconnect them to uninstall.. never will I use Snap again.

That hints at my underlying point.. separating the OS from applications for backup is the only sure fire way to ensure an always operable OS. In my opinion, the OS and GUI settings should remain completely separate from any applications which are the most likely offenders of interrupting basic user operation. Similar to the Flatpak sandboxing posture. So why not allow for the installations to be separated by partitions as well.

This is the portable applications philosophy. And it comes with many benefits as well, such as centralized and de-duplicated storage of applications for many different OSes. I think flatpaks can achieve closer adoption of this, by creating 'adapter' 'core' packages, similar I think to what the dotnet packages do on Windows. Of course, I moved to Linux primarily because of it's lighterweight to be used in VMs because I grew exhausted with full reinstalls after some application I trialed changed the user operation in a way that was irreparable, and the base install is over 30GBs now. Moving those application specific packages off to another partition is the way for a reliable and restorable VM system. Flatpak could be front and center with that by making those 'core' packages freer to be placed outside of the OS. 

The symlinking of /var/lib/flatpak already works well. I'm suggesting to further move and system 'strings attached' out of the system, and then to make separate flatpak 'cores' per OS. It would be great if that core could simply reside on the separate applications partition and be able to engage with different OSes depending on which OS called the application. AppImages can even work crossplatform on Windows with a Linux subsystem. I'm assuming flatpaks could to, but integration to centralized applications would be complex at the moment. Think of how Kubernetes provisions VMs and microservices separately. They are all single images that can connect to each other. Yes, they need to be compatible, but interoperability is becoming more and more standardized, and adapter images less required.

I'm going to continue using AppImages, even though they are bulky - likely because each one of them has to carry around a very redundant 'core' that works on multiple systems, and even though I have no love for the readability of the AppImageHub download file version names and no CLI package manager. But I would love to see Flatpak come up with a rival and better solution, because it seems so close to being there already.

**Comment by smcv** [comment]
Flatpak itself - what you are referring to as "Flatpak core" - is designed to be part of the operating system, in the same way that the Linux kernel, glibc, a desktop environment, X11/Wayland, and OpenGL/Vulkan drivers are designed to be part of the operating system. This is analogous to the way something like Docker or Kubernetes provisions each container separately, but you still have to install Docker or Kubernetes as part of your operating system - you can't use Docker to install Docker!

To get good integration with your desktop environment, Flatpak relies on services like xdg-desktop-portal, which also need to be part of your desktop environment - in particular, xdg-desktop-portal-gtk backends need to talk directly to components of your desktop environment, so they need to be compatible with the specific version of your desktop environment that you're running.

**Comment by auwsom** [comment]
@smcv Thanks, good answer.. and analogy.. I would just suggest to detach as much as possible from the OS. 

You're right Kubernetes is an orchestration system. But then AppImages run completely self contained. Using them as an example, I guess what I'm proposing is to break off what ever part of the AppImage is the adapter for different OSes into a separate ochestrator that is self-contained and independent from the OS, which must me possible given that the AppImage itself can be. Then each application file is much smaller, and the 'core' adapter is centralized, but still separable from the OS. Again, flatpak seems to be close to this, even if not designed specifically for it, but the way things are going with 'thin provisioning', I'm suggesting it would be really great for Flatpak to think about this model. For, example, there might be a way to then convert a Flatpak application to a docker container, and vice versa. Thus really improving interoperability.

**Comment by orowith2os** [comment]
It could be possible to modify Flatpak to allow running a bundle inside of a runtime without installing, in which case the only requirements on a system would be a runtime, and some runtime checks for extensions if needed. It's not quite like AppImage, but it could be a decent alternative, though may be out of scope for the Flatpak project.

You'd need to modify Flatpak to ensure the configs are carried across systems, however, such as having a runnable flatpak bundle and essentially a portable `~/.var/app/APPID`. You'd have to transfer those *two* bits, instead of just one, but it has its own benefits.

I may be repeating what some others have said, but this is how I'm thinking it would work. It would also have some basic dependencies on the system, which should be there anyways: flatpak, the base runtimes (gnome, KDE, freedesktop), the kernel, portals, and the display server.

**Comment by burtvadr** [comment]
https://github.com/burtvadr/Portable-Flatpak/

I'm working on a Flatpak specification that allows Flatpak to be made portable , This is not the most elegant solution, but it already works for me. I can format my PC and keep my portable flatpak on another drive/partition.

its very rough and probably is not useful for everyone

**Comment by maverick74** [comment]
**WOW!!!**

Very very interesting!!! Would be awesome to have something like this but "officially"!

**Comment by burtvadr** [comment]
> **WOW!!!**
> 
> Very very interesting!!! Would be awesome to have something like this but "officially"!

i totally agree, That’s why I tried keeping the portable concept separate from the Symbolic links-based approach.
